### PR TITLE
promtail: add support for TLS/mTLS in syslog receiver

### LINKED ISSUE #3898: promtail: add support for (m)TLS in syslog receiver
**Is your feature request related to a problem? Please describe.**

Most syslog servers (at least rsyslog and syslog-ng) support TLS for log forwarding. It would be great if promtrail could be secured using TLS and even with mTLS to validate that the client is trusted to send logs.

**Describe the solution you'd like**

A `tls_config` section could be added to `scrape_config[*].syslog` using the existing `tls_config` object with the following semantic:

* `ca_file`: CA used to validate client certificate (enables mTLS if set)
* `cert_file`: the certificate or chain sent by the server
* `key_file`: the private key used by the server
* `server_name`: unused
* `insecure_skip_verify`: unsused

**Describe alternatives you've considered**

This could also be achieved by putting a TCP proxy with (m)TLS support in front of promtail, but the feature seems easy enough to implement that adding a proxy seems way more complex.

**Additional context**

Related to #1275.

See [A step by step guide to mTLS in Go](https://venilnoronha.io/a-step-by-step-guide-to-mtls-in-go) for details on how to implement this feature.

This patch should be enough to make it work (not tested, not even compiled):

```
diff --git a/pkg/promtail/scrapeconfig/scrapeconfig.go b/pkg/promtail/scrapeconfig/scrapeconfig.go
index 881bc1506..907f44314 100644
--- a/pkg/promtail/scrapeconfig/scrapeconfig.go
+++ b/pkg/promtail/scrapeconfig/scrapeconfig.go
@@ -5,6 +5,7 @@ import (
 	"reflect"
 	"time"
 
+	promconfig "github.com/prometheus/common/config"
 	"github.com/prometheus/common/model"
 	"github.com/prometheus/prometheus/discovery"
 	"github.com/prometheus/prometheus/discovery/azure"
@@ -164,6 +165,8 @@ type SyslogTargetConfig struct {
 	// UseIncomingTimestamp sets the timestamp to the incoming syslog mesages
 	// timestamp if it's set.
 	UseIncomingTimestamp bool `yaml:"use_incoming_timestamp"`
+
+	TLSConfig promconfig.TLSConfig `yaml:"tls_config,omitempty"`
 }
 
 // WindowsEventsTargetConfig describes a scrape config that listen for windows event logs.
diff --git a/pkg/promtail/targets/syslog/syslogtarget.go b/pkg/promtail/targets/syslog/syslogtarget.go
index 4023b8bfc..90281d394 100644
--- a/pkg/promtail/targets/syslog/syslogtarget.go
+++ b/pkg/promtail/targets/syslog/syslogtarget.go
@@ -2,6 +2,8 @@ package syslog
 
 import (
 	"context"
+	"crypto/tls"
+	"crypto/x509"
 	"errors"
 	"fmt"
 	"net"
@@ -83,8 +85,39 @@ func NewSyslogTarget(
 	return t, err
 }
 
+func (t *SyslogTarget) newTLSConfig() (tls.Config, error) {
+	if len(t.config.TLSConfig.CertFile) < 0 || len(t.config.TLSConfig.KeyFile) < 0 {
+		return nil, fmt.Errorf("Certificate and key files are required")
+	}
+
+	certs, err := tls.LoadX509KeyPair(t.config.TLSConfig.CertFile, t.config.TLSConfig.KeyFile)
+	if err != nil {
+		return nil, fmt.Errorf("unable to load server certificate or key: %w", err)
+	}
+
+	tlsConfig := &tls.Config{
+		Certificates:       []tls.Certificate{cert},
+		InsecureSkipVerify: cfg.InsecureSkipVerify,
+	}
+
+	if len(t.config.TLSConfig.CAFile) > 0 {
+		clientCA, err := ioutil.ReadFile(t.config.TLSConfig.CAFile)
+		if err != nil {
+			return nil, fmt.Errorf("unable to load client CA: %w", err)
+		}
+
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(caCert)
+
+		tlsConfig.ClientCAs = caCertPool
+		tlsConfig.ClientAuth = tls.RequireAndVerifyClientCert
+	}
+
+	return tlsConfig, nil
+}
+
 func (t *SyslogTarget) run() error {
-	l, err := net.Listen("tcp", t.config.ListenAddress)
+	l, err := tls.Listen("tcp", t.config.ListenAddress, t.newTLSConfig())
 	l = conntrack.NewListener(l, conntrack.TrackWithName("syslog_target/"+t.config.ListenAddress))
 	if err != nil {
 		return fmt.Errorf("error setting up syslog target %w", err)
```

### PR DESCRIPTION
**Pull request by johanfleury**
**What this PR does / why we need it**:

Add support for TLS and mTLS in promtail's syslog receiver.

**Which issue(s) this PR fixes**:

Fixes #3898 

**Special notes for your reviewer**:

Go is not my main language (and I don’t really like it) so there might be some coding-style issue or some bad patterns.

I successfully tested this change using rsyslog and I got the following results:

* TLS disabled on both sides: **works as before**
* TLS enabled on both sides: **works as before (with TCP sessions being encrypted)**
* TLS enabled but client sends plain text messages: fails with `level=warn ts=2021-06-25T17:34:09.985034229Z caller=syslogtarget.go:207 msg="error initializing syslog stream" err="tls: first record does not look like a TLS handshake"`
* TLS enabled on both sides, with client certificate verification: **works as before (with TCP sessions being encrypted)**
* TLS enabled on both sides, with client certificate verification, but client doesn’t send a certificate: fails with `level=warn ts=2021-06-25T17:28:22.787523967Z caller=syslogtarget.go:207 msg="error initializing syslog stream" err="tls: client didn't provide a certificate"`

**Checklist**:

- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

File: clients/pkg/promtail/targets/syslog/syslogtarget.go
Comment (johanfleury) [inline clients/pkg/promtail/targets/syslog/syslogtarget.go:98]: Not sure if this is the best way to check if `tls_config` is set.

File: clients/pkg/promtail/targets/syslog/syslogtarget.go
Comment (johanfleury) [inline clients/pkg/promtail/targets/syslog/syslogtarget.go:117]: This is inspired by [NewTLSConfig from github.com/prometheus/common/config](https://pkg.go.dev/github.com/prometheus/common/config#NewTLSConfig).

File: docs/sources/clients/promtail/configuration.md
Comment (KMiller-Grafana) [inline docs/sources/clients/promtail/configuration.md:752]:
```suggestion
  # CA certificate used to validate client certificate. Enables client certificate verification when specified.
```

**Review by KMiller-Grafana — CHANGES_REQUESTED** [review]
Spelling correction is needed in the Docs file. Otherwise, the docs portion of the PR looks good to me.

File: docs/sources/clients/promtail/configuration.md
Comment (johanfleury) [reply docs/sources/clients/promtail/configuration.md]: Fixed, thank you :)

**Review by cyriltovena — APPROVED** [review]
LGTM

**Comment by cyriltovena** [comment]
@KMiller-Grafana I'll need your approval to merge this one please, thanks !

**Review by KMiller-Grafana — APPROVED** [review]
Thank you for the doc change.  Docs LGTM now.

**Comment by johanfleury** [comment]
Can we merge this if everybody’s happy with it?

**Comment by johanfleury** [comment]
These bots are awful.

**Pull request merged**
